src/com/facebook/buck/maven/Resolver.java

```diff
@@ -98,75 +120,115 @@
+      IOException {
+    final RepositorySystem repoSys = locator.getService(RepositorySystem.class);
+    final RepositorySystemSession session = newSession(repoSys);
 
     ImmutableMap<String, Artifact> knownDeps = getRunTimeTransitiveDeps(
         repoSys,
         session,
-        mavenCoords);
+        specifiedDependencies.values());
 
     // We now have the complete set of dependencies. Build the graph of dependencies. We'd like
     // aether to do this for us, but it doesn't preserve the complete dependency information we need
     // to accurately construct build files.
-    MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
+    final MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
 
     // Now we have the graph, grab the sources and jars for each dependency, as well as the relevant
     // checksums (which are download by default. Yay!)
+    ImmutableSetMultimap<Path, Prebuilt> downloadedArtifacts =
+        downloadArtifacts(repoSys, session, graph);
 
-    Map<Path, SortedSet<Prebuilt>> buckFiles = new HashMap<>();
+    createBuckFiles(downloadedArtifacts);
+  }
 
-    for (Artifact artifact : graph.getNodes()) {
-      List<String> extraVisibility = specifiedArtifacts.contains(artifact.toString()) ?
-        this.visibility : ImmutableList.<String>of();
-      downloadArtifact(artifact, repoSys, session, buckFiles, graph, extraVisibility);
-    }
+  private ImmutableSetMultimap<Path, Prebuilt> downloadArtifacts(
+      final RepositorySystem repoSys,
+      final RepositorySystemSession session,
+      final MutableDirectedGraph<Artifact> graph) throws ExecutionException, InterruptedException {
+    ListeningExecutorService exec = MoreExecutors.listeningDecorator(
+        Executors.newFixedThreadPool(
+            Runtime.getRuntime().availableProcessors(),
+            new MostExecutors.NamedThreadFactory("artifact download")));
+
+    @SuppressWarnings("unchecked")
+    List<ListenableFuture<Map.Entry<Path, Prebuilt>>> results =
+        (List<ListenableFuture<Map.Entry<Path, Prebuilt>>>) (List<?>)
+        exec.invokeAll(FluentIterable.from(graph.getNodes())
+            .transform(new Function<Artifact, Callable<Map.Entry<Path, Prebuilt>>>() {
+              @Override
+              public Callable<Map.Entry<Path, Prebuilt>> apply(final Artifact artifact) {
+                return new Callable<Map.Entry<Path, Prebuilt>>() {
+                  @Override
+                  public Map.Entry<Path, Prebuilt> call() throws Exception {
+                    return downloadArtifact(artifact, repoSys, session, graph);
+                  }
+                };
+              }
+            })
+            .toList());
 
-    createBuckFiles(buckFiles);
+    try {
+      return ImmutableSetMultimap.<Path, Prebuilt>builder()
+          .orderValuesBy(Ordering.natural())
+          .putAll(Futures.allAsList(results).get())
```

Comment:
I think you're misreading the code a bit. 

The code looks like:
```java
List<ListenableFuture<...>> results = exec.invokeAll(List<Callable<...>>);
```

`invokeAll` is a `List<Callable<...>> => List<ListenableFuture<...>>` and `Futures.allAsList` just transforms `List<ListenableFuture<...>> => ListenableFuture<List<...>>` so I could call `.get`. This is all just a very annoying way of doing a parallel map - I'm more than happy to change it if there's a better solution.